processor/tailsamplingprocessor/internal/sampling/composite.go

```diff
@@ -0,0 +1,162 @@
+// limitations under the License.
+
+package sampling
+
+import (
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+)
+
+type subpolicy struct {
+	// the subpolicy evaluator
+	evaluator PolicyEvaluator
+
+	// spans per second allocated to each subpolicy
+	allocatedSPS int64
+
+	// spans per second that each subpolicy sampled in this period
+	sampledSPS int64
+}
+
+// Composite evaluator and its internal data
+type Composite struct {
+	// the subpolicy evaluators
+	subpolicies []*subpolicy
+
+	// maximum total spans per second that must be sampled
+	maxTotalSPS int64
+
+	// current unix timestamp second
+	currentSecond int64
+
+	// The time provider (can be different from clock for testing purposes)
+	timeProvider TimeProvider
+
+	logger *zap.Logger
+}
+
+var _ PolicyEvaluator = (*Composite)(nil)
+
+// SubPolicyEvalParams defines the evaluator and max rate for a sub-policy
+type SubPolicyEvalParams struct {
+	Evaluator         PolicyEvaluator
+	MaxSpansPerSecond int64
+}
+
+// NewComposite creates a policy evaluator that samples all subpolicies.
+func NewComposite(
+	logger *zap.Logger,
+	maxTotalSpansPerSecond int64,
+	subPolicyParams []SubPolicyEvalParams,
+	timeProvider TimeProvider,
+) PolicyEvaluator {
+
+	subpolicies := []*subpolicy{}
+
+	for i := 0; i < len(subPolicyParams); i++ {
+		sub := &subpolicy{}
+		sub.evaluator = subPolicyParams[i].Evaluator
+		sub.allocatedSPS = subPolicyParams[i].MaxSpansPerSecond
+
+		// We are just starting, so there is no previous input, set it to 0
+		sub.sampledSPS = 0
+
+		subpolicies = append(subpolicies, sub)
+	}
+
+	return &Composite{
+		maxTotalSPS:  maxTotalSpansPerSecond,
+		subpolicies:  subpolicies,
+		timeProvider: timeProvider,
+		logger:       logger,
+	}
+}
+
+// OnLateArrivingSpans notifies the evaluator that the given list of spans arrived
+// after the sampling decision was already taken for the trace.
+// This gives the evaluator a chance to log any message/metrics and/or update any
+// related internal state.
+func (c *Composite) OnLateArrivingSpans(Decision, []*pdata.Span) error {
+	c.logger.Debug("Spans are arriving late, decision is already made!!!")
+	return nil
+}
+
+// Evaluate looks at the trace data and returns a corresponding SamplingDecision.
+func (c *Composite) Evaluate(traceID pdata.TraceID, trace *TraceData) (Decision, error) {
+	// Rate limiting works by counting spans that are sampled during each 1 second
+	// time period. Until the total number of spans during a particular second
+	// exceeds the allocated number of spans-per-second the traces are sampled,
+	// once the limit is exceeded the traces are no longer sampled. The counter
+	// restarts at the beginning of each second.
+	// Current counters and rate limits are kept separately for each subpolicy.
+
+	currSecond := c.timeProvider.getCurSecond()
+	if c.currentSecond != currSecond {
+		// This is a new second
+		c.currentSecond = currSecond
+		// Reset counters
+		for i := range c.subpolicies {
```

Comment:
Any reasons not to use the stdlib rate limiter? https://pkg.go.dev/golang.org/x/time/rate